druntime/src/core/internal/array/capacity.d

```diff
@@ -9,62 +9,150 @@
 */
 module core.internal.array.capacity;
 
-// HACK: `nothrow` and `pure` is faked.
-private extern (C) void[] _d_arraysetlengthT(const TypeInfo ti, size_t newlength, void[]* p) nothrow pure;
-private extern (C) void[] _d_arraysetlengthiT(const TypeInfo ti, size_t newlength, void[]* p) nothrow pure;
-
 /*
- * This template is needed because there need to be a `_d_arraysetlengthTTrace!Tarr` instance for every
- * `_d_arraysetlengthT!Tarr`. By wrapping both of these functions inside of this template we force the
- * compiler to create a instance of both function for every type that is used.
+ * Fully templated implementation of `_d_arraysetlengthT`, removing reliance on `TypeInfo`.
+ * The old `_d_arraysetlengthT` and `_d_arraysetlengthiT` functions have been removed.
+ */
+
+/**
+ * Resize a dynamic array by setting the `.length` property.
+ *
+ * Newly created elements are initialized to their default value.
+ *
+ * This function is **now fully templated**, eliminating the need for TypeInfo-based
+ * `_d_arraysetlengthT` and `_d_arraysetlengthiT`, while efficiently handling both
+ * zero-initialized and custom-initialized types.
+ *
+ * ---
+ * ## Example Usage:
+ * ```d
+ * void main()
+ * {
+ *     int[] a = [1, 2];
+ *     a.length = 3; // gets lowered to `_d_arraysetlengthT!(int[])(a, 3)`
+ * }
+ * ```
+ * ---
+ *
+ * - Uses a **templated approach** to minimize `TypeInfo` dependencies.
+ * - Follows `_d_newarrayU` for allocation and initialization.
+ * - Handles memory allocation and resizing safely, ensuring correct initialization.
  */
 
-/// Implementation of `_d_arraysetlengthT` and `_d_arraysetlengthTTrace`
-template _d_arraysetlengthTImpl(Tarr : T[], T)
+/// Complete templated implementation of `_d_arraysetlengthT` and its GC profiling variant `_d_arraysetlengthTTrace`
+
+size_t _d_arraysetlengthT(Tarr : T[], T)(
```

Comment:
Consider updating the lowering to use this signature:
```d
size_t _d_arraysetlengthT(Tarr : T[], T)(return scope ref Tarr arr, size_t newlength, bool isMutable)
```

Then the compiler should unqualify `arr` itself before calling the hook in `expressionsem.d` like so [1]. The point of this would be to only emit one template instance regardless of type qualifiers. So instead of using `is (T = ...)` below you'd just use `isMutable`). This might not work, so experiment with it first. It's best to apply this suggestion last, when everything else works.

[1] https://github.com/dlang/dmd/blob/8c4a95f083f9ba65941630cd5e2498a54e652db5/compiler/src/dmd/expressionsem.d#L5425-L5432